src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs

```diff
@@ -30,5 +33,43 @@ public static ImmutableArray<Document> GetVisibleDocuments(this IDocumentTrackin
                       .WhereNotNull()
                       .Distinct()
                       .ToImmutableArray();
+
+        public static Task DelayIfNonVisibleAsync(this IDocumentTrackingService service, Document document, TimeSpan timeSpan, CancellationToken cancellationToken)
+            => DelayIfNonVisibleAsync(service, ImmutableArray.Create(document), timeSpan, cancellationToken);
+
+        public static async Task DelayIfNonVisibleAsync(this IDocumentTrackingService service, ImmutableArray<Document> documents, TimeSpan timeSpan, CancellationToken cancellationToken)
+        {
+            // Only add a delay if we have access to a service that will tell us when the documents become visible or not.
+            if (!service.SupportsDocumentTracking)
+                return;
+
+            // if any of the docs are visible, then no delay.
+            var visibleDocuments = service.GetVisibleDocuments();
+            var anyVisible = documents.Select(d => d.Id).Intersect(visibleDocuments).Any();
+            if (anyVisible)
+                return;
+
+            // Listen to when the active document changed so that we startup work on a document once it becomes visible.
+
+            var delayTask = Task.Delay(timeSpan, cancellationToken);
+            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+            service.ActiveDocumentChanged += OnActiveDocumentChanged;
```

Comment:
Should this be hooked up before doing the `GetVisibleDocuments` call above, just in case?

Reply:
yes it should!  good catch!